src/gui/EditWidgetIcons.cpp

```diff
@@ -158,107 +153,79 @@ void EditWidgetIcons::setUrl(const QString& url)
-            tempurl.setScheme("http");
-            m_url = tempurl.url();
-            tempurl.setPath("/favicon.ico");
-            fetchFavicon(tempurl);
+            emit messageEditEntry(tr("Unable to fetch favicon."), MessageWidget::Error);
         }
-    });
+    } else {
+        emit messageEditEntry(tr("Unable to fetch favicon.") + "\n" +
+                              tr("Hint: You can enable Google as a fallback under Tools>Settings>Security"),
+                              MessageWidget::Error);
+    }
 
-    m_ui->faviconButton->setDisabled(true);
+    m_ui->faviconButton->setDisabled(false);
+#endif
 }
 
-void EditWidgetIcons::fetchFaviconFromGoogle(const QString& domain)
+#ifdef WITH_XC_NETWORKING
+size_t writeCurlResponse(char* ptr, size_t size, size_t nmemb, void* data)
 {
-    if (config()->get("security/IconDownloadFallbackToGoogle", false).toBool() && m_fallbackToGoogle) {
-        resetFaviconDownload();
-        m_fallbackToGoogle = false;
-        QUrl faviconUrl = QUrl("https://www.google.com/s2/favicons");
-        faviconUrl.setQuery("domain=" + QUrl::toPercentEncoding(domain));
-        fetchFavicon(faviconUrl);
-    } else {
-        resetFaviconDownload();
-        emit messageEditEntry(tr("Unable to fetch favicon."), MessageWidget::Error);
-    }
+    QByteArray* response = static_cast<QByteArray*>(data);
+    size_t realsize = size * nmemb;
+    response->append(ptr, realsize);
+    return realsize;
 }
 
-void EditWidgetIcons::resetFaviconDownload(bool clearRedirect)
+QImage EditWidgetIcons::fetchFavicon(const QUrl& url)
 {
-    if (clearRedirect) {
-        m_redirectUrl.clear();
-        m_redirectCount = 0;
-    }
+    QImage image;
+    CURL* curl = curl_easy_init();
+    if (curl != nullptr) {
+        QByteArray imagedata;
+        QByteArray baUrl = url.url().toLatin1();
+
+        curl_easy_setopt(curl, CURLOPT_URL, baUrl.data());
+        curl_easy_setopt(curl, CURLOPT_MAXREDIRS, 5L);
+        curl_easy_setopt(curl, CURLOPT_REDIR_PROTOCOLS, CURLPROTO_HTTP | CURLPROTO_HTTPS);
+        curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
+        curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT, 5L);
+        curl_easy_setopt(curl, CURLOPT_FAILONERROR, 1L);
+        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &imagedata);
+        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, &writeCurlResponse);
+
+        // Perform the request in another thread
+        QFuture<CURLcode> future = QtConcurrent::run([curl]() {
+            return curl_easy_perform(curl);
+        });
+
+        QEventLoop loop;
+        QFutureWatcher<CURLcode> watcher;
+        connect(&watcher, SIGNAL(finished()), &loop, SLOT(quit()));
+        watcher.setFuture(future);
+        loop.exec();
+
+        curl_easy_cleanup(curl);
 
-    if (nullptr != m_httpClient) {
-        m_httpClient->deleteLater();
-        m_httpClient = nullptr;
+        if (future.result() == CURLE_OK) {
+            image.loadFromData(imagedata);
+        }
```

Comment:
Maybe show an error message otherwise?

Reply:
This is a worker function, the error messages are handled in the parent caller (downloadFavicon())